static/panes/output.ts

```diff
@@ -0,0 +1,273 @@
+    fontScale: FontScale;
+    wrapButton: JQuery<HTMLElement>;
+    normalAnsiToHtml: AnsiToHtml.Filter;
+    errorAnsiToHtml: AnsiToHtml.Filter;
+    wrapTitle: string;
+    options: Toggles;
+    constructor(hub: Hub, container: Container, state: OutputState & PaneState) {
+        // canonicalize state
+        if((state as any).compiler) state.id = (state as any).compiler;
+        if((state as any).editor) state.editorid = (state as any).editor;
+        if((state as any).tree) state.treeid = (state as any).tree;
+        super(hub, container, state);
+        this.hub = hub;
+        this.contentRoot = this.domRoot.find('.content');
+        this.optionsToolbar = this.domRoot.find('.options-toolbar');
+        this.fontScale = new FontScale(this.domRoot, state, '.content');
+        this.fontScale.on('change', this.updateState.bind(this));
+        this.normalAnsiToHtml = makeAnsiToHtml();
+        this.errorAnsiToHtml = makeAnsiToHtml('red');
+        this.eventHub.emit('outputOpened', this.compilerInfo.compilerId);
+        this.onOptionsChange();
+    }
+
+    override getInitialHTML(): string {
+        return $('#compiler-output').html();
+    }
+
+    override registerOpeningAnalyticsEvent() {
+        ga.proxy('send', {
+            hitType: 'event',
+            eventCategory: 'OpenViewPane',
+            eventAction: 'Output',
+        });
+    }
+
+    override registerButtons(state: OutputState & PaneState) {
+        this.wrapButton = this.domRoot.find('.wrap-lines');
+        this.wrapTitle = this.wrapButton.prop('title');
+        // TODO: Would be nice to be able to get rid of this cast
+        this.options = new Toggles(this.domRoot.find('.options'), state as unknown as Record<string, boolean>);
+    }
+
+    override registerCallbacks() {
+        this.options.on('change', this.onOptionsChange.bind(this));
+        this.eventHub.on('compiling', this.onCompiling, this);
+    }
+
+    onOptionsChange() {
+        const options = this.getEffectiveOptions();
+        this.contentRoot.toggleClass('wrap', options.wrap);
+        this.wrapButton.prop('title', '[' + (options.wrap ? 'ON' : 'OFF') + '] ' + this.wrapTitle);
+        this.updateState();
+    }
+
+    getEffectiveOptions() {
+        return this.options.get();
+    }
+
+    override resize() {
+        const rootHeight = this.domRoot.height();
+        const toolbarHeight = this.optionsToolbar.height();
+        if(rootHeight && toolbarHeight) {
+            this.contentRoot.height(rootHeight - toolbarHeight - 5);
+        }
+    }
+
+    override getCurrentState() {
+        const parent = super.getCurrentState();
+        const options = this.getEffectiveOptions();
+        const state = {
+            wrap: options.wrap,
+            ...parent,
+        };
+        this.fontScale.addState(state);
+        return state as any;
+    }
+
+    addOutputLines(result) {
+        for(const obj of (result.stdout || []).concat(result.stderr || [])) {
```

Comment:
in a compiled language i would heavily question this line of code. I'm sure it doesn't matter, but can you const var this list a line before the for loop?

Reply:
Quick clarification, are you suggesting pulling the full `(result.stdout || []).concat(result.stderr || [])` part or both `result.std___ || []` parts?

Reply:
I think it'd be a little more clear if both || []'s were pulled out

Reply:
Will hold off merging until this has been replied to - I can easily imagine such code came from my original version; and we can almost certainly do better, but I wouldn't gate @jeremy-rifkin 's change on this, necessarily.

Reply:
I'll go ahead and pull both out, @matsjla agreed it would look a little more readable